pkg/channel/message_receiver.go

```diff
@@ -185,7 +186,31 @@ func (r *MessageReceiver) ServeHTTP(response nethttp.ResponseWriter, request *ne
 		r.reporter.ReportEventCount(&args, nethttp.StatusBadRequest)
 		return
 	}
-	err = r.receiverFunc(request.Context(), channel, message, []binding.Transformer{}, utils.PassThroughHeaders(request.Header))
+
+	bufferedMessage, err := buffering.CopyMessage(request.Context(), message)
+	if err != nil {
+		r.logger.Warn("Cannot buffer cloudevent message", zap.Error(err))
+		response.WriteHeader(nethttp.StatusBadRequest)
+		_ = r.reporter.ReportEventCount(&args, nethttp.StatusBadRequest)
+		return
+	}
+
+	event, err := binding.ToEvent(request.Context(), bufferedMessage)
+	if err != nil {
+		r.logger.Warn("failed to extract event from request", zap.Error(err))
+		response.WriteHeader(nethttp.StatusBadRequest)
+		_ = r.reporter.ReportEventCount(&args, nethttp.StatusBadRequest)
+		return
+	}
+
+	// run validation for the extracted event
+	if err := event.Validate(); err != nil {
+		r.logger.Warn("failed to validate extracted event", zap.Error(err))
+		response.WriteHeader(nethttp.StatusBadRequest)
+		return
+	}
```

Comment:
could we move this section into a more encapsulated function?

Reply:
what should that function return/do?

Reply:
The binding, validation?

It's ok to inline and grow the functions, but I sometimes feel our funcs get longer and longer.

Reply:
We can do in a separate PR, if we want to (of course not a blocker) :sweat_smile: